src/main/java/org/jenkinsci/plugins/nomad/MigrationHelper.java

```diff
@@ -0,0 +1,328 @@
+import org.jenkinsci.plugins.nomad.Api.PortGroup;
+import org.jenkinsci.plugins.nomad.Api.Resource;
+import org.jenkinsci.plugins.nomad.Api.RestartPolicy;
+import org.jenkinsci.plugins.nomad.Api.Task;
+import org.jenkinsci.plugins.nomad.Api.TaskGroup;
+import org.jenkinsci.plugins.nomad.Api.Vault;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonObject;
+
+import hudson.Util;
+import hudson.util.Secret;
+import jenkins.model.Jenkins;
+
+/**
+ * Helps with migrating plugin configurations from an older version to the current one.
+ */
+public class MigrationHelper {
+
+    /**
+     * Migrates a given {@link NomadCloud} (created with plugin version < 0.8.0) to the current version. If no migration is necessary
+     * then the given {@link NomadCloud} remains unchanged.
+     * @since 0.8.0
+     */
+    public static void migrate(NomadCloud cloud) {
+        migrateJenkinsUrl(cloud);
+        migrateWorkerUrl(cloud);
+        migrateTemplates(cloud);
+    }
+
+    private static void migrateTemplates(NomadCloud cloud) {
+        String jenkinsUrl = getFieldValue(cloud, "jenkinsUrl");
+        String jenkinsTunnel = getFieldValue(cloud, "jenkinsTunnel");
+        String workerUrl = getFieldValue(cloud, "workerUrl");
+
+        cloud.getTemplates().forEach(template -> {
+            migrateDriver(template);
+            migrateJobTemplate(template, jenkinsUrl, jenkinsTunnel, workerUrl);
+        });
+    }
+
+    private static void migrateWorkerUrl(NomadCloud cloud) {
+        String workerUrl = getFieldValue(cloud, "workerUrl");
+
+        if (StringUtils.isEmpty(workerUrl)) {
+            String jenkinsUrl = getFieldValue(cloud, "jenkinsUrl");
+            setFieldValue(cloud, "workerUrl", Util.ensureEndsWith(jenkinsUrl, "/") + "jnlpJars/slave.jar");
+        }
+    }
+
+    private static void migrateJenkinsUrl(NomadCloud cloud) {
+        String jenkinsUrl = getFieldValue(cloud, "jenkinsUrl");
+
+        if (StringUtils.isEmpty(jenkinsUrl)) {
+            setFieldValue(cloud, "jenkinsUrl", Jenkins.get().getRootUrl());
+        }
+    }
+
+    private static void migrateDriver(NomadWorkerTemplate template) {
+        String driver = getFieldValue(template, "driver");
+
+        if (StringUtils.isEmpty(driver)) {
+            Boolean useRawExec = getFieldValue(template, "useRawExec");
+
+            if (Boolean.TRUE.equals(useRawExec)) {
+                driver  = "raw_exec";
+            } else {
+                String image = getFieldValue(template, "image");
+                driver = StringUtils.isEmpty(image) ? "java" : "docker";
+            }
+
+            setFieldValue(template, "driver", driver);
+        }
+    }
+
+    private static void migrateJobTemplate(NomadWorkerTemplate template, String jenkinsUrl, String jenkinsTunnel, String workerUrl) {
+        String jobTemplate = getFieldValue(template, "jobTemplate");
+
+        if (StringUtils.isEmpty(jobTemplate)) {
```

Comment:
Maybe we should just do a `null` check here (and maybe for the other migrations as well). Assumption: new fields have the default value which is in this case `null`.

Reply:
This looks enough to me, which problems do you foresee?

Reply:
No problem at all I guess. It's just that the criteria for the migration is not 100% correct (null vs "").